File: 0141-linked-list-cycle/0141-linked-list-cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
    slowP, fastP := head, head // slow pointer, fast pointer are start at same node
    for fastP != nil && fastP.Next != nil {
        slowP = slowP.Next
        fastP = fastP.Next.Next
        if slowP == fastP {
            return true
        }
    }
    return false
}

/**
 * The O(n) time and space complexity is using map or set to store the visited node,
 * and check if the next node was visited, return true, otheriwise, return false (at the end).
 * 
 * But there is O(n) time and O(1) space complexity, I am going to use two pointer with diff pace,
 * so if the linked-list is cyclic, the pointers will meet.
 */

File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *curr1 = head, *curr2 = head;

        while (curr1 && curr2 && curr2->next) {
            if (curr1 == curr2->next) {
                return true;
            }

            curr1 = curr1->next;
            curr2 = curr2->next->next;
        }

        return false;
    }
};